ThunderCloud/TSCAppDelegate.swift

```diff
@@ -0,0 +1,196 @@
+//
+//  TSCAppDelegate.swift
+//  ThunderCloud
+//
+//  Created by Simon Mitchell on 24/08/2017.
+//  Copyright © 2017 threesidedcube. All rights reserved.
+//
+
+import UIKit
+import UserNotifications
+import ThunderRequest
+
+@UIApplicationMain
+/// A root app delegate which sets up your window and push notifications e.t.c.
+open class TSCAppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
+
+	/// The main window of the app
+	public var window: UIWindow?
+	
+	/// Whether to show push notifications when the app is in the foreground
+	public var foregroundNotificationOptions: UNNotificationPresentationOptions? = [.alert, .badge, .sound]
+	
+	open func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
+		
+		UNUserNotificationCenter.current().delegate = self
+		
+		window = UIWindow(frame: UIScreen.main.bounds)
+		window?.backgroundColor = .white
+		
+		window?.rootViewController = TSCAppViewController()
+		window?.makeKeyAndVisible()
+		
+		setupSharedUserAgent()
+		
+		DeveloperModeController.shared.installDeveloperMode(toWindow: window!, currentTheme: TSCTheme())
+		
+		// Register errors
+		TSCErrorRecoveryAttempter.registerOverrideDescription(
+			"Failed to load page".localised(with: "_STREAMINGPAGE_FAILED_TITLE"),
+			recoverySuggestion: "We were unable to find the page for this notification".localised(with: "_STREAMINGPAGE_FAILED_RECOVERYSUGGESTION"),
+			forDomain: "ThunderCloud.streamingError",
+			code: 1)
+				
+		return true
+	}
+	
+	//MARK: -
+	//MARK: - Push notifications
+	//MARK: -
+	
+	open func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+		TSCStormNotificationHelper.registerPushToken(deviceToken)
+	}
+
+	open func handleNotificationResponse(for notification: UNNotification, response: UNNotificationResponse?, fromLaunch: Bool) -> Bool {
+		
+		// Make sure the action wasn't dismissing the notification
+		guard response?.actionIdentifier != UNNotificationDismissActionIdentifier else {
+			return false
+		}
+		
+		guard let payload = notification.request.content.userInfo["payload"] as? [AnyHashable : Any] else { return false }
+		guard let url = payload["url"] as? String else { return false }
```

Comment:
Could combine this with the guard above, as they're both returning false :)

Reply:
Actually, if this is being recalled once per second while we're presenting the view controller, we could perhaps check `window?.rootViewController?.presentedViewController != nil` first before unwrapping the payload, as otherwise it isn't being used. What do you think? :)

Reply:
That's fine with me, I will try and move these changes over to the `feature/swift` branch as well 👍

Reply:
Actually scrap that, sorry @ryanbourneuk the reason these are above the check for presented view controller is because we don't even need to re-call once per second if there is no payload url. Will move them into one guard statement but they need to stay above the rest of the method

Reply:
Ah yes! That's very true indeed! As you were :)